src/fsharp/FSharp.Core/iseq.fs

```diff
@@ -0,0 +1,1345 @@
+                        this.Result <- LanguagePrimitives.DivideByInt<'T> this.Result this.State 
+                    override this.OnDispose () = () })
+
+        [<CompiledName "AverageBy">]
+        let inline averageBy (f:'T->'U) (source:ISeq<'T>) =
+            source.Fold (fun _ ->
+                upcast { new FolderWithPostProcessing<'T,'U,int>(LanguagePrimitives.GenericZero,0) with
+                    override this.ProcessNext value =
+                        this.Result <- Checked.(+) this.Result (f value)
+                        this.State <- this.State + 1
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ =
+                        if this.State = 0 then
+                            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                        this.Result <- LanguagePrimitives.DivideByInt<'U> this.Result this.State
+                    override this.OnDispose () = () })
+
+        [<CompiledName "ExactlyOne">]
+        let exactlyOne (source:ISeq<'T>) : 'T =
+            source.Fold (fun pipeIdx ->
+                upcast { new FolderWithPostProcessing<'T,'T,Values<bool, bool>>(Unchecked.defaultof<'T>, Values<bool,bool>(true, false)) with
+                    override this.ProcessNext value =
+                        if this.State._1 then
+                            this.State._1 <- false
+                            this.Result <- value
+                        else
+                            this.State._2 <- true
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ =
+                        if this.State._1 then
+                            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                        elif this.State._2 then
+                            invalidArg "source" (SR.GetString SR.inputSequenceTooLong)
+                    override this.OnDispose () = () })
+
+        [<CompiledName "Fold">]
+        let inline fold<'T,'State> (f:'State->'T->'State) (seed:'State) (source:ISeq<'T>) : 'State =
+            source.Fold (fun _ ->
+                { new Folder<'T,'State>(seed) with
+                    override this.ProcessNext value =
+                        this.Result <- f this.Result value
+                        Unchecked.defaultof<_> (* return value unused in Fold context *) })
+
+        [<CompiledName "Fold2">]
+        let inline fold2<'T1,'T2,'State> (folder:'State->'T1->'T2->'State) (state:'State) (source1:ISeq<'T1>) (source2: ISeq<'T2>) =
+            source1.Fold (fun pipeIdx ->
+                upcast { new FolderWithPostProcessing<_,'State,IEnumerator<'T2>>(state,source2.GetEnumerator()) with
```

Comment:
I'm fairly confident that in all my implementations I handle - in the ISeq.Fold implementations this is handled by manual discipline, and example being the finally block in [this code](https://github.com/manofstick/visualfsharp/blob/ded9bb71682328b92a22f765471c2c72d59f118f/src/fsharp/FSharp.Core/seqcore.fs#L324-L334). Here the ChainDispose will call up through the base class of [FolderWithPostProcessing](https://github.com/manofstick/visualfsharp/blob/ded9bb71682328b92a22f765471c2c72d59f118f/src/fsharp/FSharp.Core/iseq.fs#L60-L72) which does an exception protected call to the OnDipose on the fold2 implementation.

Now as with any code, I may have a bug in there, but I think I have made it _fairly_ easy to get right - or at least as easy as I could. There were a number of tests that did check for correct disposing (not added by me, they were there checking the previous Seq implementations)